Introduce Lotus::Model::Error

karimmtarek opened the pull request:
- Added Error class
- Required Error class in model.rb and null_adapter.rb
- Replaced all ::StandardError with Lotus::Model::Error
- Added tests

jodosha commented on lib/lotus/model.rb:
```diff
@@ -16,7 +17,7 @@ module Model
     # @since 0.1.0
     #
     # @see Lotus::Repository.update
```
@karimmtarek Instead of requiring a separated file with that exception, can you please define it here? Thank you!

jodosha commented:
@karimmtarek This looks good, but for extra safety, can you please add `test/error_test.rb`? Where we unit test that `Lotus::Model::Error` inherits from `::StandardError`. Then add one unit test for each custom exception that we have in the framework, and assert that they inherit from `Lotus::Model::Error`.

Apart from this improvement, the PR looks good. Thank you very much! :+1:

karimmtarek commented:
@jodosha I moved the `Error` class inside `model.rb` though I think now the `Error` class seems not to load properly in some `modules`, @pascalbetz helped my out and looked at the code and proposed couple of suggestions(please refer to the chat room for the complete transcript):
- Move `Error` class to it's own file and require it when needed.
- Some restructuring will be required if we'll keep the `Error` class inside `model.rb`

Other than this, I added couple of tests in the error_test.rb please let me know if this is what you had in mind.

joneslee85 commented on lib/lotus/model.rb:
```diff
@@ -25,13 +28,13 @@ class NonPersistedEntityError < ::StandardError
     #
     # @see Lotus::Configuration#mapping
-    class InvalidMappingError < ::StandardError
+    class InvalidMappingError < Error
```
:clap: nice work
while you are on it, can we add test for every exception class?

karimmtarek replied:
I added couple of test [there](https://github.com/karimmtarek/model/blob/introduce-lotus-model-error/test/error_test.rb), I just need to know if I'm testing the right thing, can you please look into it and advice?

joneslee85 replied:
I left a comment there, btw more integration (if possible) test is welcome

joneslee85 commented on test/error_test.rb:
```diff
@@ -0,0 +1,11 @@
+    Lotus::Model::Error.superclass.must_be StandardError
+  end
+
+  it 'parent to all custom exception' do
```
I don't think this test adds more value because you have already tested it in `test/integration/repository_test.rb`

karimmtarek commented:
@jodosha the tests I added are failing, is it because I need to require a file(s) I'm not sure, can you please look into it and advise, thanks.

karimmtarek commented:
@jodosha please review this when ever you have some time, thanks.

jodosha commented:
@karimmtarek This looks good. Thank you! :+1:
